homeassistant/components/gree/__init__.py

```diff
@@ -21,52 +31,61 @@ async def async_setup(hass: HomeAssistant, config: dict):
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     """Set up Gree Climate from a config entry."""
-    devices = []
+    gree_discovery = DiscoveryService(hass)
+    hass.data[DATA_DISCOVERY_SERVICE] = gree_discovery
 
-    # First we'll grab as many devices as we can find on the network
-    # it's necessary to bind static devices anyway
-    _LOGGER.debug("Scanning network for Gree devices")
+    @callback
+    def shutdown_event(_: Event) -> None:
+        if hass.data[DOMAIN].get(DATA_DISCOVERY_INTERVAL) is not None:
+            hass.data[DOMAIN][DATA_DISCOVERY_INTERVAL]()
+            hass.data[DOMAIN].pop(DATA_DISCOVERY_INTERVAL)
+
+        if hass.data.get(DATA_DISCOVERY_SERVICE) is not None:
+            del hass.data[DATA_DISCOVERY_SERVICE]
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown_event)
 
-    for device_info in await DeviceHelper.find_devices():
-        try:
-            device = await DeviceHelper.try_bind_device(device_info)
-        except CannotConnect:
-            _LOGGER.error("Unable to bind to gree device: %s", device_info)
-            continue
-
-        _LOGGER.debug(
-            "Adding Gree device at %s:%i (%s)",
-            device.device_info.ip,
-            device.device_info.port,
-            device.device_info.name,
-        )
-        devices.append(device)
-
-    coordinators = [DeviceDataUpdateCoordinator(hass, d) for d in devices]
-    await asyncio.gather(*[x.async_refresh() for x in coordinators])
-
-    hass.data[DOMAIN][COORDINATOR] = coordinators
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(entry, CLIMATE_DOMAIN)
     )
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(entry, SWITCH_DOMAIN)
     )
 
+    _LOGGER.debug("Scanning network for Gree devices")
+    await gree_discovery.discovery.scan()
+
+    async def _async_scan_update(now):
```

Comment:
What's the `now` parameter for? If intentionally unused, maybe name it `_` or `_now`.

Reply:
Parameter appears to be require by `async_track_time_interval` for the callback. I'll rename it to `_`